clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1050,6 +1062,81 @@ private void close(Duration timeout, boolean swallowException) {
         }
     }
 
+    /**
+     * Prior to closing the network thread, we need to make sure the following operations happen in the right sequence:
+     * 1. autocommit offsets
+     * 2. revoke all partitions
+     */
+    private void prepareShutdown(final Timer timer, final AtomicReference<Throwable> firstException) {
+        if (!groupMetadata.isPresent())
+            return;
+        maybeAutoCommitSync(timer, firstException);
+        timer.update();
+        waitOnEventCompletion(new ConsumerCloseApplicationEvent(ConsumerCloseApplicationEvent.Task.COMMIT, timer.remainingMs()), timer, firstException);
+        maybeInvokeCommitCallbacks();
+        maybeRevokePartitions(timer, firstException);
+        waitOnEventCompletion(new ConsumerCloseApplicationEvent(ConsumerCloseApplicationEvent.Task.LEAVE_GROUP, timer.remainingMs()), timer, firstException);
+    }
+
+    private void waitOnEventCompletion(final ConsumerCloseApplicationEvent event,
+                                       final Timer timer,
+                                       final AtomicReference<Throwable> firstException) {
+        try {
+            applicationEventHandler.addAndGet(event, timer);
+        } catch (TimeoutException e) {
```

Comment:
We don't really throw timeout exceptions during closing because if user tries to close with 0 duration then all ops will be timedout.  The current implementation just polls, but since we cannot directly polls the client, we need to either wait till the future is completed or times out and keep going.

Reply:
@kirktrue and I discussed the potential tasks for dealing with zero timeout.  This needs to be examined perhaps after the preview. So we will spin off a jira ticket for this specific issue.